### PR TITLE
[IMP] mail,  fleet, hr_recruitment : integrate fleet and documents

### PR DESCRIPTION
**Pull request by ppr-odoo**
- improve the document flow for fleet vehicle.

  Purpose of the task is to add stat button on vehicle form view
  for attachment and on click of that button open the attachment
  kanban view. from this button user can create an attachment.
  
  Also some code refactor for hr_recruitment. removed unused field
  document_ids and rename the count field as attachment_count instead
  of document_count(due to conflict with name in document.mixin).

-  filter attachment and disable upload in chatter

    Currently, we can not filter attachment in chatter and we can not
    disable the 'add an attachment' button in chatter.
    Purpose of the task is to disable the upload button and filter the
    chatter attachments so user can upload the document using stat button
    of document app.
    
    So in this commit, Added the options 'disable_upload' and
    'attachment_domain' in chatter field so user can filter attachment
    and disable the upload from bottom chatter of the form view.
    For ex - 
               options="{'disable_upload': [('documents_fleet_settings', '=', True)],
                                    'attachment_domain': [('documents_ids', '=', False)]}"
  
closes odoo/odoo#78592
task-2144780

### CODE REVIEW

**Review by seb-odoo — CHANGES_REQUESTED** [review]
Hello, here's a first check for the mail part.

See comments below, also I'm not sure to understand what is this specific domain override?

File: addons/mail/static/src/components/chatter_container/chatter_container.js [hunk 1]
```diff
@@ -100,6 +100,14 @@ Object.assign(ChatterContainer, {
             type: Boolean,
             optional: true,
         },
+        disableAttachmentUpload: {
+            type: Boolean,
+            optional: true,
+        },
+        attachmentDomain: {
```
[the remark this answers is not in the document]
Comment (ppr-odoo) [reply addons/mail/static/src/components/chatter_container/chatter_container.js:107]: currently, domain directly pass in search_read.

If we pass a boolean in props, it has to be passed from context to search_read and search_read has to be override. the record is filtered from that context. Thanks

[the remark this answers is not in the document]
Comment (ppr-odoo) [reply addons/mail/static/src/components/chatter_container/chatter_container.js:107]: @seb-odoo

[the remark this answers is not in the document]
Comment (ppr-odoo) [reply addons/mail/static/src/components/chatter_container/chatter_container.js:107]: @seb-odoo currently, domain directly pass in search_read.

If we pass a boolean in props, it has to be passed from context to search_read and search_read has to be override. the record is filtered from that context. Thanks

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/components/chatter_container/chatter_container.js:107]: No change to `search_read`, I mean "convert" in JS before calling the method (in `fetchAttachments`).

File: addons/mail/models/mail_thread.py [hunk 2]
```diff
@@ -3142,3 +3142,7 @@ def _get_mail_redirect_suggested_company(self):
         if 'company_id' in self:
             return self.company_id
         return False
+
+    @api.model
+    def _get_attachment_domain(self, thread, data):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:3147]: I don't think thread mixin should have a method just for a given domain, that is tweaked for some strange reason in documents. This should be done as a tool function in controllers directly, this is not a business method :) .

Comment (ppr-odoo) [reply addons/mail/models/mail_thread.py]: @tde-banana-odoo  Done.Thanks

File: addons/mail/static/src/components/attachment_box/attachment_box.xml [hunk 3]
```diff
@@ -18,7 +18,7 @@
                             record="attachmentBoxView.attachmentList"
                         />
                     </t>
-                    <button class="o_AttachmentBox_buttonAdd btn btn-link" type="button" t-on-click="attachmentBoxView.onClickAddAttachment">
+                    <button class="o_AttachmentBox_buttonAdd btn btn-link" type="button" t-on-click="attachmentBoxView.onClickAddAttachment" t-if="!attachmentBoxView.chatter.hasAttachmentUpload">
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/components/attachment_box/attachment_box.xml:21]: Also, the name of field is the opposite of what it actually does.
Instead swap the default value (`false` => `true`), and here remove the `!`.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/components/attachment_box/attachment_box.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ppr-odoo) [reply addons/mail/static/src/components/attachment_box/attachment_box.xml]: > Also, the name of field is the opposite of what it actually does. Instead swap the default value (`false` => `true`), and here remove the `!`.

Done

File: addons/mail/static/src/components/attachment_box/attachment_box.js [hunk 4]
```diff
@@ -29,7 +29,13 @@ export class AttachmentBox extends Component {
...
 
 Object.assign(AttachmentBox, {
-    props: { record: Object },
+    props: {
+        record: Object ,
+        hasAttachmentUpload: {
+            type: Boolean,
+            optional: true,
+        },
+    },
```
[the remark this answers is not in the document]
Comment (ppr-odoo) [reply addons/mail/static/src/components/attachment_box/attachment_box.js]: > `AttachmentBox(Component)/hasAttachmentUpload` is not used, so revert this diff.

Done.Thanks

File: addons/hr_recruitment/models/hr_job.py [hunk 5]
```diff
@@ -124,8 +123,7 @@ def _compute_document_ids(self):
                 result[attachment.res_id] |= attachment
 
         for job in self:
-            job.document_ids = result.get(job.id, False)
-            job.documents_count = len(job.document_ids)
+            job.attachment_count = len(result[job.id]) if result else False
```
Comment (Williambraecky) [inline addons/hr_recruitment/models/hr_job.py:126]: Why these changes?

**Review by Williambraecky — COMMENTED** [review]
Hello, thanks for your work

I have asked the current PO about the behaviour you introduced for the chatter as it seems highly unnecessary, it seems logical for a fleet user to have access to all attachments.
The issue is more within the fleet app rather than the document handling.

File: addons/mail/static/src/widgets/form_renderer/form_renderer.js [hunk 6]
```diff
@@ -92,6 +93,7 @@ FormRenderer.include({
      * @returns {Object}
      */
     _makeChatterContainerProps() {
+        const hasAttachmentUpload = new Domain(this.chatterFields.hasAttachmentUpload).compute(this.state.evalContext);
```
Comment (seb-odoo) [inline addons/mail/static/src/widgets/form_renderer/form_renderer.js:96]: Do we want this? And if so, do we really want this to be a chatter option? Isn't it strictly model dependent?

File: addons/mail/static/src/models/thread.js [hunk 7]
```diff
@@ -528,11 +540,7 @@ registerModel({
                 suggestedRecipients: suggestedRecipientsData,
             } = await this.messaging.rpc({
                 route: '/mail/thread/data',
-                params: {
-                    request_list: [...requestSet],
-                    thread_id: this.id,
-                    thread_model: this.model,
-                },
+                params: this._getThreadParam(requestSet),
```
Comment (seb-odoo) [inline addons/mail/static/src/models/thread.js:543]: I don't like this change, this API should be kept simple.

File: addons/mail/models/mail_thread.py [hunk 8]
```diff
@@ -3133,11 +3133,11 @@ def _get_mail_redirect_suggested_company(self):
             return self.company_id
         return False
 
-    def _get_mail_thread_data_attachments(self):
+    def _get_mail_thread_data_attachments(self, data):
         self.ensure_one()
         return self.env['ir.attachment'].search([('res_id', '=', self.id), ('res_model', '=', self._name)], order='id desc')
 
-    def _get_mail_thread_data(self, request_list):
+    def _get_mail_thread_data(self, request_list, data):
```
Comment (seb-odoo) [inline addons/mail/models/mail_thread.py:3140]: I don't like the `data` param in any of these methods.

**Review by seb-odoo — CHANGES_REQUESTED** [review]
Hello, it seems there are a lot of changes that we don't really want, as @Williambraecky already posted above I will not comment into too much details, but please let me know when this PR reaches a final stage and I'll have a look again.

**Comment by C3POdoo** [comment]
Dear @ppr-odoo,

Thank you for your contribution but we are closing it due to inactivity.
We only support the last 3 stable versions so no longer accepts patches into this branch.

We apology if we could not look at your request in time.
If the contribution still makes sense, please let us know and do not hesitate to recreate one for the recent versions. We will try to check it as soon as possible.

_This is an automated message._
